### PR TITLE
Fix corner cases in Bash runfiles library

### LINKED ISSUE #16933: Bazel 6.0.0rc4 breaks Bazel RBE tests
This was first detected in the downstream test for Bazel 6.0.0rc4: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2761#0184d3b5-3cf0-410a-984d-2fdfdf78a49b

And verified in https://buildkite.com/bazel/bazel-bazel/builds/21471#0184e876-6406-4909-9a81-e8dc957b471f

[Integration tests with test args](https://cs.opensource.google/search?q=%22args%20%3D%20%5B%22$(JAVABASE)%22%5D%22&sq=&ss=bazel) are failing remotely with test logs look like:
```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //src/test/shell/bazel:bazel_example_test
-----------------------------------------------------------------------------
WARNING: Passing test names in arguments (--test_arg) is deprecated, please use --test_filter='/usr/lib/jvm/java-11-openjdk-amd64' instead.
WARNING: Arguments do not specify tests!
INFO[bazel_example_test 2022-12-06 17:28:07 (+0000)] bazel binary is at /b/f/w/bazel-out/k8-fastbuild/bin/src/test/shell/bazel/bazel_example_test.runfiles/io_bazel/src/test/shell/bin
INFO[bazel_example_test 2022-12-06 17:28:07 (+0000)] setting up client in /b/f/w/_tmp/55e5a95f98bdfb926d4a13d2054933c5/workspace
```

### PR DESCRIPTION
**Pull request by fmeum**
Due to tests not actually verifying the exit code of rlocation calls, a few special cases were not handled correctly. In particular, absolute paths still went through regular rlocation lookup and could fail there even after the absolute path had been printed to stdout.
    
Also adds a missing newline in the (very rare) case that a manifest entry is found, but doesn't point to an existing file (e.g. if it is an unresolved symlink), and `rlocation` is not used with command substitution, but some other mechanism that doesn't strip trailing newlines.
    
The tests now assert the expected exit code (== 0 or != 0).
    
Fixes #16933

### CODE REVIEW

**Comment by fmeum** [comment]
@Wyverald Could you review?

FYI @meteorcloudy

**Comment by fmeum** [comment]
I don't understand the CI failure yet. `android_tools.tar` isn't found, but it seems that is expected.

**Comment by gregestren** [comment]
Looks okay to me, pending failures. @ahumesky any Android comments?

**Comment by ted-xie** [comment]
> android_tools.tar isn't found, but it seems that is expected.

Seems fishy... why shouldn't it be found?

I have a couple questions about this PR:
* Do we have to keep the `set -x` and `env = {RUNFILES_LIB_DEBUG}` lines in the diff? Seems like those are for debug purposes mostly. I am mildly concerned that leaving these in will inflate build and test logs globally for all Bazel users. I could also be misunderstanding the intent here.
* The boilerplate runfiles incantation is repeated a bunch of times across the bazel repo, and probably in user projects too. Will we have to update these incantations if you update runfiles.bash?

File: src/test/shell/bazel/android/aidl_integration_test.sh [hunk 1]
```diff
@@ -14,6 +14,8 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+set -x
```
Comment (ted-xie) [inline src/test/shell/bazel/android/aidl_integration_test.sh:17]: Please take these debug commands out if they're only for debug.

**Review by ted-xie — CHANGES_REQUESTED** [review]

**Comment by fmeum** [comment]
> > android_tools.tar isn't found, but it seems that is expected.
> 
> Seems fishy... why shouldn't it be found?

It looks like only a few tests declare `android_tools.tar` as a data dependency, but many tests call through to a helper function that locates it. Not sure whether that's correct, but I will update the PR to remain compatible with its previous behavior in this situation.

> * Do we have to keep the `set -x` and `env = {RUNFILES_LIB_DEBUG}` lines in the diff? Seems like those are for debug purposes mostly. I am mildly concerned that leaving these in will inflate build and test logs globally for all Bazel users. I could also be misunderstanding the intent here.

These are going away, I am just (ab)using CI to debug issues on Windows.

> * The boilerplate runfiles incantation is repeated a bunch of times across the bazel repo, and probably in user projects too. Will we have to update these incantations if you update runfiles.bash?

No, the incantation has remained compatible with the new version 3 of the runfiles library.

File: src/test/shell/bazel/android/aidl_integration_test.sh (hunk 1 above)
Comment (fmeum) [reply src/test/shell/bazel/android/aidl_integration_test.sh]: Done.

**Comment by meteorcloudy** [comment]
Downstream test for this change: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2774

**Comment by ted-xie** [comment]
@meteorcloudy is the original author of the runfiles macro so I think he's the most qualified to give the final review here. No concerns from the Android side for me, as long as all the existing tests pass in presubmit.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by Wyverald** [comment]
Downstream test results look good: https://buildkite.com/bazel/bazel-at-head-plus-downstream/builds/2775#_

**Review by meteorcloudy — APPROVED** [review]
